src/resty/gcp/init.lua

```diff
@@ -51,50 +51,93 @@ FindApis = function(apiClass, methods, curr)
     return methods
 end
 
+local function template_expansion(str, params)
+    return (string.gsub(str, "{(.-)}", function (key)
+        return ngx.escape_uri(params[key])
+    end))
+end
+
+local function build_request(accesstoken, apiDetail, params, requestBody, opt)
+    local req = {
+        method = apiDetail.httpMethod,
+        headers = {
+            ["Authorization"] = "Bearer " .. accesstoken.token
+        },
+        body = requestBody,
+        ssl_verify = false
+    }
+
+    -- simple APIs
+    if apiDetail.flatPath then
+        local path = template_expansion(apiDetail.flatPath, params)
+        return path, req
+    end
+
+    local path = template_expansion(apiDetail.path, params)
+    local query = {}
+
+    for k, v in pairs(params) do
+        local param = apiDetail.parameters[k]
+        if not param then
+            error("invalid parameter: " .. k)
+        end
+        local location = param.location
+        if location == "query" then
+            query[k] = v
+            -- skip paths as they are already handled
+        end
+    end
+
+    if opt and opt.media then
+        if not apiDetail.supportsMediaDownload then
+            error("API does not supported media download")
+        end
+        query.alt = "media"
+    end
+
+    if next(query) then
+        path = path .. "?" .. ngx.encode_args(query)
+    end
+
+    return path, req
+end
+
 local BuildMethods = function(methods)
     local baseUrl = methods.baseUrl
     local services = {}
     for k, v in pairs(methods) do
         if type(v) == "table" then
             services[k] = {}
             for serviceName, apiDetail in pairs(v) do
-                services[k][serviceName] = function(accesstoken, params, requestBody)
+                services[k][serviceName] = function(accesstoken, params, requestBody, opt)
                     if (not params) then
-                        return
+                        error("params is required")
```

Comment:
breaking change?

Reply:
Actually, this repo is "under early development" and I think it's acceptable. This change added missing error handling that I consider necessary.

Reply:
yes - thanks for adding this error handling. Please make sure you test it with GCP vault backend. It used those generated API JSON stubs present under /api

Reply:
> This additional `opt` argument - what is it for? Have you tested that after adding it - those api stubs generating correctly?

I don't have much background knowledge, but it seems that some of the APIs support an option, to get the media itself or its meta-data